README.rst
DC Brick
========

This repository contains the firmware source code and the hardware design
files. The documentation generator configs can be found at
https://github.com/Tinkerforge/generators

Repository Content
------------------

software/:
 * examples/: Examples for all supported languages
 * build/: Compiled files
 * src/: Source code of firmware
 * Makefile: Makefile to build project

hardware/:
 * Contains KiCad project files and additionally schematics as PDF

datasheets/:
 * Contains datasheets for sensors and complex ICs that are used

Hardware
--------

The hardware is designed with the open source EDA Suite KiCad
(http://www.kicad-pcb.org). Before you are able to open the files,
you have to install the Tinkerforge kicad-libraries
(https://github.com/Tinkerforge/kicad-libraries). You can either clone
them directly in hardware/ or clone them in a separate folder and
symlink them into hardware/
(ln -s kicad_path/kicad-libraries project_path/hardware). After that you
can open the .pro file in hardware/ with KiCad and from there view and
modify the schematics and the PCB layout.

Software
--------

If you want to do your own Brick/Bricklet firmware development we highly
recommend that you use our build environment setup script and read the
tutorial (https://www.tinkerforge.com/en/doc/Tutorials/Tutorial_Build_Environment/Tutorial.html).

To compile the C code we recommend you to install the newest GNU Arm Embedded
Toolchain (https://launchpad.net/gcc-arm-embedded/+download).
You also need to install bricklib (https://github.com/Tinkerforge/bricklib).

You can either clone it directly in software/src/ or clone it in a
separate folder and symlink it into software/src/
(ln -s bricklib_path/bricklib project_path/software/src/). Finally make sure to
have CMake installed (http://www.cmake.org/cmake/resources/software.html).

After that you can build the firmware by invoking make in software/.
The firmware (.bin) can then be found in software/build/ and uploaded
with brickv (click button "Flashing" on start screen).


software/changelog
2011-10-06: 1.0.0 (c009c89)
- Initial version

2012-02-15: 1.1.0 (f33ee6a)
- Add support for tick types

2012-03-05: 1.1.1 (ce4afa6)
- Fix bug with timeouts on Windows (through bricklib)

2012-03-11: 1.1.2 (a0a10b7)
- Fix bug with ADC calibration readout (through bricklib)

2012-06-30: 1.1.3 (89d1ed1)
- Add get_chip_temperature and reset function

2012-07-02: 1.1.4 (4b14a66)
- Add USB hotplug support (through bricklib)

2012-09-20: 1.1.5 (9370625)
- Set interrupt pins to default (otherwise they can interfere with WIFI)

2012-10-26: 1.1.6 (66ae792)
- Change Bricklet Context size from 250 to 256 (through bricklib)

2013-01-17: 2.0.0 (c6e2870)
- Support for Protocol 2.0

2013-04-11: 2.0.1 (0eb1fdb)
- Add watchdog timer (restart after 16s of "deadlock")

2013-06-27: 2.0.2 (d3f0801)
- Use tick_task to detect USB connect/disconnect instead of interrupt

2014-07-07: 2.0.3 (63b83bf)
- Remove USB hotplug support (through bricklib)

2014-11-25: 2.3.0 (030fbc0)
- Add support for new SPI protocol

2016-04-04: 2.3.1 (e45ed60)
- Make setting velocity to 0 work while motor is disabled

2016-08-29: 2.3.2 (af79780)
- Make sure to use all chunks when writing Bricklet plugin (through bricklib)

2017-05-11: 2.3.3 (6f39bee)
- Add support for co-mcu Bricklets

2017-06-28: 2.3.4 (d0b36fce)
- Fix incompatibility between co-mcu support and LED Strip Bricklet (through bricklib)

2017-12-13: 2.3.5 (3b204424)
- Make sure Brick does not restart on suspend/resume (through bricklib)
- Refactor initial enumeration handling, fix double-enumeration bugs (through bricklib)

2018-01-11: 2.3.6 (783e4fdc)
- Fix message duplication bug (through bricklib)

2018-04-27: 2.3.7 (2b54410b)
- Fix potential deadlock between usb stack and comcu enumeration (through bricklib)

2018-07-31: 2.3.8 (f24db991)
- Add support for Isolator Bricklet (through bricklib)
- Improve Co-MCU Bricklet error handling (through bricklib)

2019-02-27: 2.3.9 (68bd0de)
- Changes through bricklib:
  - Fix Bricklet port miss-identification
  - Increase SPI stack timeout (for RED Brick)
  - Fix bugs in set/get_spitfp_* functions

2020-05-14: 2.3.10 (89631c1)
- Fix get drive mode response length


software/src/communication.h
/* dc-brick
 * Copyright (C) 2010-2012 Olaf Lüke <olaf@tinkerforge.com>
 *
 * communication.h: Implementation of DC-Brick specific messages
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */


#ifndef COMMUNICATION_H
#define COMMUNICATION_H

#include <stdint.h>
#include <stdbool.h>

#include "bricklib/com/com_common.h"

#define FID_SET_VELOCITY 1
#define FID_GET_VELOCITY 2
#define FID_GET_CURRENT_VELOCITY 3
#define FID_SET_ACCELERATION 4
#define FID_GET_ACCELERATION 5
#define FID_SET_PWM_FREQUENCY 6
#define FID_GET_PWM_FREQUENCY 7
#define FID_FULL_BRAKE 8
#define FID_GET_STACK_INPUT_VOLTAGE 9
#define FID_GET_EXTERNAL_INPUT_VOLTAGE 10
#define FID_GET_CURRENT_CONSUMPTION 11
#define FID_ENABLE 12
#define FID_DISABLE 13
#define FID_IS_ENABLED 14
#define FID_SET_MINIMUM_VOLTAGE 15
#define FID_GET_MINIMUM_VOLTAGE 16
#define FID_SET_DRIVE_MODE 17
#define FID_GET_DRIVE_MODE 18
#define FID_SET_CURRENT_VELOCITY_PERIOD 19
#define FID_GET_CURRENT_VELOCITY_PERIOD 20
#define FID_UNDER_VOLTAGE 21
#define FID_EMERGENCY_SHUTDOWN 22
#define FID_VELOCITY_REACHED 23
#define FID_CURRENT_VELOCITY 24

#define COM_MESSAGE_USER_LAST_FID 24

#ifdef ENCODER
#define FID_ENABLE_ENCODER 25
#define FID_DISABLE_ENCODER 26
#define FID_IS_ENCODER_ENABLED 27
#define FID_GET_ENCODER_COUNT 28
#define FID_SET_ENCODER_CONFIG 29
#define FID_GET_ENCODER_CONFIG 30
#define FID_SET_ENCODER_PID_CONFIG 31
#define FID_GET_ENCODER_PID_CONFIG 32
#endif

#define COM_MESSAGES_USER \
	{FID_SET_VELOCITY, (message_handler_func_t)set_velocity}, \
	{FID_GET_VELOCITY, (message_handler_func_t)get_velocity}, \
	{FID_GET_CURRENT_VELOCITY, (message_handler_func_t)get_current_velocity}, \
	{FID_SET_ACCELERATION, (message_handler_func_t)set_acceleration}, \
	{FID_GET_ACCELERATION, (message_handler_func_t)get_acceleration}, \
	{FID_SET_PWM_FREQUENCY, (message_handler_func_t)set_pwm_frequency}, \
	{FID_GET_PWM_FREQUENCY, (message_handler_func_t)get_pwm_frequency}, \
	{FID_FULL_BRAKE, (message_handler_func_t)full_brake}, \
	{FID_GET_STACK_INPUT_VOLTAGE, (message_handler_func_t)get_stack_input_voltage}, \
	{FID_GET_EXTERNAL_INPUT_VOLTAGE, (message_handler_func_t)get_external_input_voltage}, \
	{FID_GET_CURRENT_CONSUMPTION, (message_handler_func_t)get_current_consumption}, \
	{FID_ENABLE, (message_handler_func_t)enable}, \
	{FID_DISABLE, (message_handler_func_t)disable}, \
	{FID_IS_ENABLED, (message_handler_func_t)is_enabled}, \
	{FID_SET_MINIMUM_VOLTAGE, (message_handler_func_t)set_minimum_voltage}, \
	{FID_GET_MINIMUM_VOLTAGE, (message_handler_func_t)get_minimum_voltage}, \
	{FID_SET_DRIVE_MODE, (message_handler_func_t)set_drive_mode}, \
	{FID_GET_DRIVE_MODE, (message_handler_func_t)get_drive_mode}, \
	{FID_SET_CURRENT_VELOCITY_PERIOD, (message_handler_func_t)set_current_velocity_period}, \
	{FID_GET_CURRENT_VELOCITY_PERIOD, (message_handler_func_t)get_current_velocity_period}, \
	{FID_UNDER_VOLTAGE, (message_handler_func_t)NULL}, \
	{FID_EMERGENCY_SHUTDOWN, (message_handler_func_t)NULL}, \
	{FID_VELOCITY_REACHED, (message_handler_func_t)NULL}, \
	{FID_CURRENT_VELOCITY, (message_handler_func_t)NULL},

#ifdef ENCODER
	{FID_ENABLE_ENCODER, (message_handler_func_t)enable_encoder}, \
	{FID_DISABLE_ENCODER, (message_handler_func_t)disable_encoder}, \
	{FID_IS_ENCODER_ENABLED, (message_handler_func_t)is_encoder_enabled}, \
	{FID_GET_ENCODER_COUNT, (message_handler_func_t)get_encoder_count}, \
	{FID_SET_ENCODER_CONFIG, (message_handler_func_t)set_encoder_config}, \
	{FID_GET_ENCODER_CONFIG, (message_handler_func_t)get_encoder_config}, \
	{FID_SET_ENCODER_PID_CONFIG, (message_handler_func_t)set_encoder_pid_config}, \
	{FID_GET_ENCODER_PID_CONFIG, (message_handler_func_t)get_encoder_pid_config},
#endif

typedef struct {
	MessageHeader header;
	int16_t velocity;
} __attribute__((__packed__)) SetVelocity;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetVelocity;

typedef struct {
	MessageHeader header;
	int16_t velocity;
} __attribute__((__packed__)) GetVelocityReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetCurrentVelocity;

typedef struct {
	MessageHeader header;
	int16_t velocity;
} __attribute__((__packed__)) GetCurrentVelocityReturn;

typedef struct {
	MessageHeader header;
	uint16_t acceleration;
} __attribute__((__packed__)) SetAcceleration;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAcceleration;

typedef struct {
	MessageHeader header;
	uint16_t acceleration;
} __attribute__((__packed__)) GetAccelerationReturn;

typedef struct {
	MessageHeader header;
	uint16_t frequency;
} __attribute__((__packed__)) SetPWMFrequency;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetPWMFrequency;

typedef struct {
	MessageHeader header;
	uint16_t frequency;
} __attribute__((__packed__)) GetPWMFrequencyReturn;


typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) FullBrake;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetStackInputVoltage;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) GetStackInputVoltageReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetExternalInputVoltage;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) GetExternalInputVoltageReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetCurrentConsumption;

typedef struct {
	MessageHeader header;
	uint16_t current;
} __attribute__((__packed__)) GetCurrentConsumptionReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) Enable;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) Disable;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) IsEnabled;

typedef struct {
	MessageHeader header;
	bool enabled;
} __attribute__((__packed__)) IsEnabledReturn;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) SetMinimumVoltage;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetMinimumVoltage;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) GetMinimumVoltageReturn;

typedef struct {
	MessageHeader header;
	uint8_t mode;
} __attribute__((__packed__)) SetDriveMode;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetDriveMode;

typedef struct {
	MessageHeader header;
	uint8_t mode;
} __attribute__((__packed__)) GetDriveModeReturn;

typedef struct {
	MessageHeader header;
	uint16_t period;
} __attribute__((__packed__)) SetCurrentVelocityPeriod;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetCurrentVelocityPeriod;

typedef struct {
	MessageHeader header;
	uint16_t period;
} __attribute__((__packed__)) GetCurrentVelocityPeriodReturn;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) UnderVoltageCallback;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) EmergencyShutdownCallback;

typedef struct {
	MessageHeader header;
	int16_t velocity;
} __attribute__((__packed__)) VelocityReachedCallback;

typedef struct {
	MessageHeader header;
	int16_t velocity;
} __attribute__((__packed__)) CurrentVelocityCallback;

#ifdef ENCODER
typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) EnableEncoder;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) DisableEncoder;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) IsEncoderEnabled;

typedef struct {
	MessageHeader header;
	bool enabled;
} __attribute__((__packed__)) IsEncoderEnabledReturn;

typedef struct {
	MessageHeader header;
	bool reset;
} __attribute__((__packed__)) GetEncoderCount;

typedef struct {
	MessageHeader header;
	int32_t count;
} __attribute__((__packed__)) GetEncoderCountReturn;

typedef struct {
	MessageHeader header;
	uint16_t counts_per_revolution;
} __attribute__((__packed__)) SetEncoderConfig;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetEncoderConfig;

typedef struct {
	MessageHeader header;
	uint16_t counts_per_revolution;
} __attribute__((__packed__)) GetEncoderConfigReturn;

typedef struct {
	MessageHeader header;
	float p;
	float i;
	float d;
	uint8_t sample_time;
} __attribute__((__packed__)) SetEncoderPIDConfig;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetEncoderPIDConfig;

typedef struct {
	MessageHeader header;
	float p;
	float i;
	float d;
	uint8_t sample_time;
} __attribute__((__packed__)) GetEncoderPIDConfigReturn;
#endif

void set_velocity(const ComType com, const SetVelocity *data);
void get_velocity(const ComType com, const GetVelocity *data);
void get_current_velocity(const ComType com, const GetCurrentVelocity *data);
void set_acceleration(const ComType com, const SetAcceleration *data);
void get_acceleration(const ComType com, const GetAcceleration *data);
void set_pwm_frequency(const ComType com, const SetPWMFrequency *data);
void get_pwm_frequency(const ComType com, const GetPWMFrequency *data);
void full_brake(const ComType com, const FullBrake *data);
void get_stack_input_voltage(const ComType com, const GetStackInputVoltage *data);
void get_external_input_voltage(const ComType com, const GetExternalInputVoltage *data);
void get_current_consumption(const ComType com, const GetCurrentConsumption *data);
void enable(const ComType com, const Enable *data);
void disable(const ComType com, const Disable *data);
void is_enabled(const ComType com, const IsEnabled *data);
void set_minimum_voltage(const ComType com, const SetMinimumVoltage *data);
void get_minimum_voltage(const ComType com, const GetMinimumVoltage *data);
void set_drive_mode(const ComType com, const SetDriveMode *data);
void get_drive_mode(const ComType com, const GetDriveMode *data);
void set_current_velocity_period(const ComType com, const SetCurrentVelocityPeriod *data);
void get_current_velocity_period(const ComType com, const GetCurrentVelocityPeriod *data);

#ifdef ENCODER
void enable_encoder(const ComType com, const EnableEncoder *data);
void disable_encoder(const ComType com, const DisableEncoder *data);
void is_encoder_enabled(const ComType com, const IsEncoderEnabled *data);
void get_encoder_count(const ComType com, const GetEncoderCount *data);
void set_encoder_config(const ComType com, const SetEncoderConfig *data);
void get_encoder_config(const ComType com, const GetEncoderConfig *data);
void set_encoder_pid_config(const ComType com, const SetEncoderPIDConfig *data);
void get_encoder_pid_config(const ComType com, const GetEncoderPIDConfig *data);
#endif

#endif
